{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "302ee60a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c56e96ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports\n",
    "import pandas as pd\n",
    "from collections import defaultdict\n",
    "from mjol.base import *\n",
    "from mjol.gan import *\n",
    "from mjol.tools import *\n",
    "import pyfastx\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8da8bc54",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_integrity(x) -> bool:\n",
    "    for aid in x.lookup:\n",
    "        for uid in x.lookup[aid]:\n",
    "            if uid not in x.features:\n",
    "                return False\n",
    "    return True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "ee5f38a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "o_fn = \"/ccb/salz4-3/hji20/chm13/chm13v2.0_RefSeq_Liftoff_v5.2.gff3\"\n",
    "o_chm13 = GAn(\n",
    "    file_name = o_fn,\n",
    "    file_fmt = 'gff'\n",
    ")\n",
    "o_chm13.build_db()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "0beac8a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n_genes = 0\n",
    "for f in o_chm13.features.values():\n",
    "    if f.feature_type == 'gene':\n",
    "        n_genes += 1\n",
    "check_integrity(o_chm13)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "07b7a6f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "nw_fn = \"/ccb/salz4-3/hji20/chm13/T2T_annotation_v5.3/output/mapped_reformatted.gff\"\n",
    "nw_chm13 = GAn(\n",
    "    file_name = nw_fn,\n",
    "    file_fmt = 'gff'\n",
    ")\n",
    "nw_chm13.build_db()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d2ada259",
   "metadata": {},
   "outputs": [],
   "source": [
    "swap_fn = \"/ccb/salz4-3/hji20/chm13/T2T_annotation_v5.3/intermediates/overlaps_8.csv\"\n",
    "swaps_df = pd.read_csv(swap_fn, index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c6e4eddb",
   "metadata": {},
   "outputs": [],
   "source": [
    "syn_df = swaps_df[swaps_df['synonym'] == True]\n",
    "non_syn_df = swaps_df[swaps_df['synonym'] == False]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "90cc22d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# variables for logging\n",
    "added_lns = []\n",
    "removed_lns = []\n",
    "actions = dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "82f9389d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_attributes(\n",
    "    starter,\n",
    "    nw_att,\n",
    "    update_rule,\n",
    "    verbose : bool = False\n",
    ") -> dict:\n",
    "    for o_k, nw_k in update_rule.items():\n",
    "        if nw_k not in nw_att:\n",
    "            if verbose:\n",
    "                print(f'WARNING : expected {nw_k} in nw_att arg, but not found')\n",
    "            continue\n",
    "        starter[o_k] = nw_att[nw_k]\n",
    "    return starter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "2ba8a0bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "cn_tbl = defaultdict(int)\n",
    "for f in o_chm13.features.values():\n",
    "    if f.feature_type == \"gene\":\n",
    "        if 'gene' in f.attributes:\n",
    "            gname = f.attributes['gene']\n",
    "        elif 'gene_name' in f.attributes:\n",
    "            gname = f.attributes['gene_name']\n",
    "        cn_tbl[gname] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "3ab18587",
   "metadata": {},
   "outputs": [],
   "source": [
    "nw_pfn = \"/ccb/salz4-3/hji20/chm13/mapped_protein.fa\"\n",
    "nw_pfa = pyfastx.Fasta(nw_pfn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "24616941",
   "metadata": {},
   "outputs": [],
   "source": [
    "vorf_tbl = dict()\n",
    "for f in nw_chm13.features.values():\n",
    "    if f.feature_type == \"mRNA\":\n",
    "        if 'valid_ORF' not in f.attributes:\n",
    "            vorf_tbl[f.aid] = 'None'\n",
    "            continue\n",
    "        if f.attributes['valid_ORF'] == 'True':\n",
    "            pseq = nw_pfa[f.aid].seq\n",
    "            # consider adding stop codon check?\n",
    "            if '.' in pseq or pseq[0] != 'M':\n",
    "                vorf_tbl[f.aid] = False\n",
    "            else:\n",
    "                vorf_tbl[f.aid] = True\n",
    "        else:\n",
    "            vorf_tbl[f.aid] = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "988ddbad",
   "metadata": {},
   "outputs": [],
   "source": [
    "sorted_syn_df = syn_df.sort_values(by='gene_add') # must operate on sorted syn_df\n",
    "sorted_non_syn_df = non_syn_df.sort_values(by='gene_add')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "1501c983",
   "metadata": {},
   "outputs": [],
   "source": [
    "# NOTE: these changes are irreversible\n",
    "\n",
    "btype_tbl = defaultdict(int)\n",
    "\n",
    "curr_cn_tbl = defaultdict(int)\n",
    "cn_ctr = 0\n",
    "prev_add_gname = None # prev added gene name\n",
    "txes_to_add = [] # tx to add manually\n",
    "genes_to_check = [] # gene to check copy number attributes\n",
    "resolved_ranges = dict()\n",
    "\n",
    "for i, row in sorted_syn_df.iterrows():\n",
    "    add_btype = row['biotype_add']\n",
    "    del_btype = row['biotype_remove']\n",
    "    btype_tbl[del_btype] += 1\n",
    "    assert add_btype == \"protein_coding\" # sanity check\n",
    "\n",
    "    o_gname = row['gene_remove']\n",
    "    o_gid = row['ID_remove']\n",
    "    o_uid = o_chm13.get_uid(o_gid)\n",
    "    og = o_chm13.get_feature(o_uid)\n",
    "\n",
    "    if cn_tbl[o_gname] > 1:\n",
    "        genes_to_check.append(o_gname)\n",
    "    \n",
    "    nw_gname = row['gene_add']\n",
    "    nw_gid = row['ID_add']\n",
    "    nw_uid = nw_chm13.get_uid(nw_gid)\n",
    "    ng = nw_chm13.get_feature(nw_uid)\n",
    "\n",
    "    # update cn_ctr\n",
    "    if prev_add_gname != nw_gname:\n",
    "        curr_cn_tbl[prev_add_gname] = cn_ctr\n",
    "        prev_add_gname = nw_gname\n",
    "        cn_ctr = 0\n",
    "    else:\n",
    "        cn_ctr += 1\n",
    "    \n",
    "    if del_btype == \"protein_coding\":\n",
    "        removed_lns.append(og.to_gff_entry(include_children=True))\n",
    "        \n",
    "        if cn_ctr > 0:\n",
    "            new_gaid = f'{ng.aid.split('_')[0]}_{cn_ctr}'\n",
    "        else:\n",
    "            new_gaid = ng.aid.split('_')[0]\n",
    "\n",
    "        update_rule = {\n",
    "            'gene_name' : 'Name',\n",
    "            'db_xref' : 'Dbxref',\n",
    "            'description' : 'description',\n",
    "            'gene' : 'gene',\n",
    "            'gene_biotype' : 'gene_biotype',\n",
    "            'coverage' : 'coverage',\n",
    "            'sequence_ID' : 'sequence_ID',\n",
    "        }\n",
    "\n",
    "        g_att = update_attributes(\n",
    "            starter = { 'ID' : new_gaid },\n",
    "            nw_att = ng.attributes,\n",
    "            update_rule = update_rule\n",
    "        )\n",
    "\n",
    "        # additional attribute updates\n",
    "        g_att['valid_ORFs'] = og.attributes['valid_ORFs']\n",
    "        g_att['extra_copy_number'] = str(cn_ctr)\n",
    "        g_att['copy_num_ID'] = f'{g_att['ID']}_{cn_ctr}' if cn_ctr == 0 else g_att['ID']\n",
    "        \n",
    "        # update gene\n",
    "        og.attributes = g_att\n",
    "        del o_chm13.lookup[og.aid]\n",
    "        og.aid = g_att['ID']\n",
    "        o_chm13.lookup[og.aid] = [og.uid]\n",
    "\n",
    "        tgt_tx = ng.children[0]\n",
    "        tgt_itrn_chain = tgt_tx.get_chain('exon')\n",
    "        add_tgt = True\n",
    "\n",
    "        for tx in og.children:\n",
    "            if tgt_itrn_chain == tx.get_chain('exon'):\n",
    "                add_tgt = False\n",
    "            \n",
    "            temp = tx.attributes\n",
    "            if cn_ctr > 0:\n",
    "                new_taid = f'{temp['ID']}_{cn_ctr}'\n",
    "            else:\n",
    "                new_taid = temp['ID']\n",
    "            \n",
    "            t_att = update_attributes(\n",
    "                starter = {\n",
    "                    'ID' : new_taid\n",
    "                },\n",
    "                nw_att = g_att,\n",
    "                update_rule = {\n",
    "                    'Parent' : 'ID',\n",
    "                    'db_xref' : 'db_xref',\n",
    "                    'gene' : 'gene',\n",
    "                    'extra_copy_number' : 'extra_copy_number'\n",
    "                }\n",
    "            )\n",
    "            \n",
    "            t_att = update_attributes(\n",
    "                starter = t_att,\n",
    "                nw_att = temp,\n",
    "                update_rule = {\n",
    "                    'tag' : 'tag',\n",
    "                    'transcript_biotype' : 'transcript_biotype',\n",
    "                    'matches_ref_protein' : 'matches_ref_protein',\n",
    "                    'valid_ORF' : 'valid_ORF'\n",
    "                }\n",
    "            )\n",
    "\n",
    "            # update tx\n",
    "            tx.attributes = t_att\n",
    "            del o_chm13.lookup[tx.aid]\n",
    "            tx.aid = new_taid\n",
    "            tx.paid = og.aid\n",
    "            o_chm13.lookup[tx.aid] = [tx.uid]\n",
    "\n",
    "            for child in tx.children:\n",
    "                temp2 = child.attributes\n",
    "                \n",
    "                c_att = update_attributes(\n",
    "                    starter = {},\n",
    "                    nw_att = t_att,\n",
    "                    update_rule = {\n",
    "                        'Parent' : 'ID',\n",
    "                        'db_xref' : 'db_xref',\n",
    "                        'gene' : 'gene',\n",
    "                        'tag' : 'tag',\n",
    "                        'transcript_biotype' : 'transcript_biotype',\n",
    "                        'extra_copy_number' : 'extra_copy_number'\n",
    "                    }\n",
    "                )\n",
    "\n",
    "                # update child\n",
    "                child.attributes = c_att\n",
    "                assert not child.aid # sanity check\n",
    "                child.paid = tx.aid\n",
    "        added_lns.append(og.to_gff_entry(include_children=True))\n",
    "        resolved_ranges[og.aid] = (og.chr, og.start, og.end)\n",
    "        actions[i] = f'RENAME;{del_btype};{o_gid};{add_btype};{og.aid}'\n",
    "    else:\n",
    "        removed_lns.append(og.to_gff_entry(include_children=True))\n",
    "        if cn_ctr > 0:\n",
    "            new_gaid = f'{ng.aid.split('_')[0]}_{cn_ctr}'\n",
    "        else:\n",
    "            new_gaid = ng.aid.split('_')[0]\n",
    "        \n",
    "        update_rule = {\n",
    "            'gene_name' : 'Name',\n",
    "            'db_xref' : 'Dbxref',\n",
    "            'description' : 'description',\n",
    "            'gene' : 'gene',\n",
    "            'gene_biotype' : 'gene_bioype',\n",
    "            'coverage' : 'coverage',\n",
    "            'sequence_ID' : 'sequence_ID'\n",
    "        }\n",
    "        \n",
    "        g_att = update_attributes(\n",
    "            starter = { 'ID' : new_gaid },\n",
    "            nw_att = ng.attributes,\n",
    "            update_rule = update_rule\n",
    "        )\n",
    "\n",
    "        g_att['extra_copy_number'] = str(cn_ctr)\n",
    "        g_att['copy_num_ID'] = f'{g_att['ID']}_{cn_ctr}' if cn_ctr == 0 else g_att['ID']\n",
    "        \n",
    "        # add gene\n",
    "        ng.attributes = g_att\n",
    "        ng.aid = new_gaid\n",
    "        o_chm13.add_feature(\n",
    "            feature = ng,\n",
    "            include_children = False\n",
    "        )\n",
    "\n",
    "        for tx in ng.children:\n",
    "            temp = tx.aid.split('_')\n",
    "            if len(temp) == 2:\n",
    "                base = tx.aid\n",
    "            else:\n",
    "                base = '_'.join(temp[:-1])\n",
    "            new_taid = f'{base}_{cn_ctr}' if cn_ctr > 0 else base\n",
    "\n",
    "            tx_att = update_attributes(\n",
    "                starter = {\n",
    "                    'ID' : new_taid,\n",
    "                    'Parent' : new_gaid,\n",
    "                    'transcript_biotype' : 'mRNA',\n",
    "                    'extra_copy_number' : str(cn_ctr)\n",
    "                },\n",
    "                nw_att = tx.attributes,\n",
    "                update_rule = {\n",
    "                    \"db_xref\" : \"Dbxref\",\n",
    "                    \"gene\" : \"gene\",\n",
    "                    \"tag\" : \"tag\",\n",
    "                    \"matches_ref_protein\" : \"matches_ref_protein\",\n",
    "                    'valid_ORF' : 'valid_ORF'\n",
    "                }\n",
    "            )\n",
    "\n",
    "            # add tx\n",
    "            tx.attributes = tx_att\n",
    "            tx.paid = new_gaid\n",
    "            tx.aid = new_taid\n",
    "            tx.feature_type = \"transcript\"\n",
    "            o_chm13.add_feature(\n",
    "                feature = tx,\n",
    "                include_children = False\n",
    "            )\n",
    "\n",
    "            for child in tx.children:\n",
    "                c_att = update_attributes(\n",
    "                    starter = {},\n",
    "                    nw_att = tx_att,\n",
    "                    update_rule = {\n",
    "                        'Parent' : 'ID',\n",
    "                        'db_xref' : 'db_xref',\n",
    "                        'gene' : 'gene',\n",
    "                        'tag' : 'tag',\n",
    "                        'transcript_biotype' : 'transcript_biotype',\n",
    "                        'extra_copy_number' : 'extra_copy_number'\n",
    "                    }\n",
    "                )\n",
    "\n",
    "                # add child\n",
    "                child.attributes = c_att\n",
    "                child.aid = None\n",
    "                child.paid = tx.aid\n",
    "                o_chm13.add_feature(\n",
    "                    feature = child,\n",
    "                    include_children = False\n",
    "                )\n",
    "        \n",
    "        # delete og and its descendants\n",
    "        o_chm13.pop_feature(og.uid, include_children=True)\n",
    "\n",
    "        resolved_ranges[ng.aid] = (ng.chr, ng.start, ng.end)\n",
    "        added_lns.append(ng.to_gff_entry(include_children=True))\n",
    "        actions[i] = f'RENAME;{del_btype};{o_gid};{add_btype};{ng.aid}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "67f566e7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "defaultdict(int,\n",
       "            {'protein_coding': 138,\n",
       "             'pseudogene': 15,\n",
       "             'lncRNA': 1,\n",
       "             'transcribed_pseudogene': 1})"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "btype_tbl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "83d17ce7",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_genes2 = 0\n",
    "for f in o_chm13.features.values():\n",
    "    if f.feature_type == 'gene':\n",
    "        n_genes2 += 1\n",
    "check_integrity(o_chm13)\n",
    "\n",
    "assert n_genes == n_genes2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "6fb486b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "rslved_by_chr = defaultdict(list)\n",
    "\n",
    "for gid in resolved_ranges:\n",
    "    chr, st, en = resolved_ranges[gid]\n",
    "    rslved_by_chr[chr].append((gid, st, en))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "2bfaee14",
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_overlapping(x_st, x_en, y_st, y_en) -> bool:\n",
    "    \"\"\"\n",
    "    x - query\n",
    "    y - reference\n",
    "    \"\"\"\n",
    "    return max(x_st, y_st) <= min(x_en, y_en)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cea46e9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# MANUAL, IGNORE, ADD, REMOVE, AS_IS logics\n",
    "\n",
    "manual_df = None\n",
    "added_tbl = defaultdict(list)\n",
    "removed_tbl = defaultdict(list)\n",
    "\n",
    "for _, sub_df in sorted_non_syn_df.groupby('gene_add'):\n",
    "    if len(sub_df) > 1:\n",
    "        if manual_df is None:\n",
    "            manual_df = sub_df\n",
    "        else:\n",
    "            manual_df = pd.concat([manual_df, sub_df], axis=0)\n",
    "        for i, row in sub_df.iterrows():\n",
    "            actions[i] = 'MANUAL'\n",
    "    else:\n",
    "        for i, row in sub_df.iterrows():\n",
    "            is_ovlp = False # skip legal overlaps, applying logics of transitivity described above\n",
    "            for r in rslved_by_chr[row['chromosome']]:\n",
    "                gid, st, en = r\n",
    "                if is_overlapping(\n",
    "                    x_st = int(row['start_remove']),\n",
    "                    x_en = int(row['end_remove']),\n",
    "                    y_st = int(st),\n",
    "                    y_en = int(en)\n",
    "                ):\n",
    "                    is_ovlp = True\n",
    "                    break\n",
    "            if is_ovlp:\n",
    "                actions[i] = f'IGNORE;{gid}'\n",
    "                break\n",
    "    \n",
    "            nw_gid = row['ID_add']\n",
    "            nw_gname = row['gene_add']\n",
    "            nw_uid = nw_chm13.get_uid(nw_gid)\n",
    "            ng = nw_chm13.get_feature(nw_uid)\n",
    "\n",
    "            if row['overlapping_in_MANE']:\n",
    "                if str(curr_cn_tbl[nw_gname]) != ng.attributes['extra_copy_number']:\n",
    "                    print(f'CRITICAL : update copy number for {nw_gid}')\n",
    "                ng.children[0].feature_type = 'transcript'\n",
    "                out = o_chm13.add_feature(\n",
    "                    feature = ng,\n",
    "                    include_children = True\n",
    "                )\n",
    "                added_lns.append(out)\n",
    "                added_tbl[nw_gname].append(nw_gid)\n",
    "                removed_lns.append('None\\n')\n",
    "                actions[i] = f'ADD;{ng.aid}'\n",
    "            else:\n",
    "                nt = ng.children[0]\n",
    "                if nt.aid not in vorf_tbl:\n",
    "                    raise KeyError()\n",
    "                if not vorf_tbl[nt.aid]:\n",
    "                    actions[i] = f'AS_IS;INVALID;{ng.aid}'\n",
    "                    continue\n",
    "                o_gname = row['gene_remove']\n",
    "                if cn_tbl[o_gname] > 1:\n",
    "                    genes_to_check.append(o_gname)\n",
    "                o_gid = row['ID_remove']\n",
    "\n",
    "                if not row['gene_remove_in_MANE']:\n",
    "                    if o_gid in o_chm13.lookup:\n",
    "                        out = o_chm13.pop_feature(\n",
    "                            uid = o_chm13.get_uid(o_gid),\n",
    "                            include_children = True\n",
    "                        )\n",
    "                        removed_lns.append(out)\n",
    "                        removed_tbl[o_gname].append(o_gid)\n",
    "                    else:\n",
    "                        removed_lns.append('None\\n')\n",
    "                    \n",
    "                    nt.feature_type = 'transcript'\n",
    "                    out = o_chm13.add_feature(\n",
    "                        feature = ng,\n",
    "                        include_children = True\n",
    "                    )\n",
    "                    added_lns.append(out)\n",
    "                    added_tbl[nw_gname].append(nw_gid)\n",
    "                    actions[i] = f'REMOVE;{row['biotype_remove']};{o_gid};ADD;{row['biotype_add']};{ng.aid}'\n",
    "                else:\n",
    "                    og = o_chm13.get_feature(o_chm13.get_uid(o_gid))\n",
    "                    winning_condition = False\n",
    "                    if og.attributes['valid_ORFs'] == '0':\n",
    "                        winning_condition = True\n",
    "                        print(f'CRITICAL : replace {og.aid} (broken) with {ng.aid} (intact)')\n",
    "                    # satisfy a strictly longer condition?\n",
    "                    if row['gene_add_protein_length'] > row['gene_remove_protein_length']:\n",
    "                        winning_condition = False\n",
    "                        print(f'CRITICAL : replace {og.aid} (shorter) with {ng.aid} (longer)')\n",
    "                    actions[i] = f'AS_IS;WIN={winning_condition}'\n",
    "                    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "b574e758",
   "metadata": {},
   "outputs": [],
   "source": [
    "# append actions to swaps_df\n",
    "assert len(actions) == len(swaps_df) # sanity check\n",
    "sorted_actions = dict(sorted(actions.items()))\n",
    "swaps_df['action'] = list(sorted_actions.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "11a95a09",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[]\n"
     ]
    }
   ],
   "source": [
    "print(txes_to_add) # no MANE tx to manually add"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "c0cfc149",
   "metadata": {},
   "outputs": [],
   "source": [
    "# write out results\n",
    "\n",
    "assert len(added_lns) == len(removed_lns)\n",
    "\n",
    "out_dir = \"/ccb/salz4-3/hji20/chm13/out\"\n",
    "swaps_df.to_csv(os.path.join(out_dir, 'edit_infos.csv'), index=False)\n",
    "\n",
    "with open(os.path.join(out_dir, 'edited_entries.txt'), 'w') as fh:\n",
    "    for i in range(len(added_lns)):\n",
    "        fh.write(f'>>>\\n{removed_lns[i]}<<<\\n{added_lns[i]}')\n",
    "\n",
    "o_chm13.to_gff3(os.path.join(out_dir, 'out.gff'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "055de86d",
   "metadata": {},
   "outputs": [],
   "source": [
    "manual_df.to_csv(os.path.join(out_dir, 'manual_edit_infos.csv'), index=False)\n",
    "\n",
    "with open(os.path.join(out_dir, 'genes_to_check.txt'), 'w') as fh:\n",
    "    fh.write('\\n'.join(genes_to_check))\n",
    "\n",
    "with open(os.path.join(out_dir, 'added_genes.csv'), 'w') as fh:\n",
    "    for gname in added_tbl:\n",
    "        for gid in added_tbl[gname]:\n",
    "            fh.write(f'{gname},{gid}\\n')\n",
    "\n",
    "with open(os.path.join(out_dir, 'removed_genes.csv'), 'w') as fh:\n",
    "    for gname in removed_tbl:\n",
    "        for gid in removed_tbl[gname]:\n",
    "            fh.write(f'{gname},{gid}\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d8dc689",
   "metadata": {},
   "source": [
    "### manual curation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "5e8c4bf9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ARB2A'}\t{'IGNORE'}\n",
      "{'CNTNAP3'}\t{'REMOVE;LOC100420438;CNTNAP3', 'REMOVE;CNTNAP3C;CNTNAP3', 'REMOVE;LOC101930090;CNTNAP3'}\n",
      "{'EIF3CL'}\t{'COMPARE', 'REMOVE;MIR6862-2;EIF3CL'}\n",
      "{'ENTREP2'}\t{'IGNORE'}\n",
      "{'FAM246B'}\t{'REMOVE;DGCR5;FAM246B', 'REMOVE;FAM246C;FAM246B'}\n",
      "{'FAM90A12'}\t{'COMPARE', 'REMOVE;FAM90A24P_3;FAM90A12_7', 'REMOVE;FAM90A24P_4;FAM90A12_6', 'REMOVE;FAM90A24P_2;FAM90A12_8'}\n",
      "{'FAM90A13'}\t{'REMOVE;FAM90A15P;FAM90A13_2', 'REMOVE;FAM90A5P;FAM90A13_1', 'COMPARE'}\n",
      "{'FAM90A24'}\t{'COMPARE'}\n",
      "{'GK3'}\t{'AS_IS;INVALID;GK3_2', 'AS_IS;INVALID;GK3_1'}\n",
      "{'GOLGA8Q'}\t{'REMOVE;LOC100288482_1;GOLGA8Q', 'REMOVE;RN7SL196P;GOLGA8Q', 'COMPARE'}\n",
      "{'IGBP1C'}\t{'IGNORE'}\n",
      "{'KCNE1'}\t{'REMOVE;LOC124900475;KCNE1', 'REMOVE;LOC105379505;KCNE1'}\n",
      "{'LOC128125814'}\t{'REMOVE;MIR616;LOC128125814', 'ADD'}\n",
      "{'MSANTD7'}\t{'REMOVE;LOC100421372;MSANTD7', 'ADD'}\n",
      "{'OR4F21'}\t{'AS_IS;INVALID;OR4F21_2', 'AS_IS;INVALID;OR4F21', 'AS_IS;INVALID;OR4F21_1', 'AS_IS;INVALID;OR4F21_3'}\n",
      "{'POTEB3'}\t{'COMPARE', 'REMOVE;RNU6-631P;POTEB3'}\n",
      "{'POTEG'}\t{'COMPARE', 'REMOVE;RNU6-1239P;POTEG', 'REMOVE;ACTBP10_1;POTEG'}\n",
      "{'RIMBP3'}\t{'COMPARE', 'REMOVE;RN7SKP63;RIMBP3'}\n",
      "{'SAXO3'}\t{'REMOVE;LOC101059948;SAXO3', 'ADD'}\n",
      "{'TBC1D3C'}\t{'AS_IS;INVALID;TBC1D3C'}\n",
      "{'TMA7B'}\t{'REMOVE;LOC112268293;TMA7B', 'ADD'}\n",
      "{'TMEM276'}\t{'IGNORE', 'REMOVE;LOC84773;TMEM276'}\n",
      "{'ZFTRAF1'}\t{'IGNORE'}\n",
      "{'ZNG1C'}\t{'AS_IS;INVALID;ZNG1C', 'IGNORE'}\n"
     ]
    }
   ],
   "source": [
    "for _, sub_df in manual_df.groupby('gene_add'):\n",
    "    actions = []\n",
    "    for i, row in sub_df.iterrows():\n",
    "        action = 'None'\n",
    "        chr = row['chromosome']\n",
    "        st = row['start_remove']\n",
    "        en = row['end_remove']\n",
    "        for r in rslved_by_chr[chr]:\n",
    "            gid, st2, en2 = r\n",
    "            if is_overlapping(\n",
    "                x_st = st,\n",
    "                x_en = en,\n",
    "                y_st = st2,\n",
    "                y_en = en2\n",
    "            ):\n",
    "                action = 'IGNORE'\n",
    "                break\n",
    "        if action == 'None':\n",
    "            if row['overlapping_in_MANE']:\n",
    "                action = 'ADD'\n",
    "            else:\n",
    "                nw_gid = row['ID_add']\n",
    "                nw_uid = nw_chm13.get_uid(nw_gid)\n",
    "                ng = nw_chm13.get_feature(nw_uid)\n",
    "                nt = ng.children[0]\n",
    "                if not vorf_tbl[nt.aid]:\n",
    "                    action = f'AS_IS;INVALID;{ng.aid}'\n",
    "                else:\n",
    "                    if not row['gene_remove_in_MANE']:\n",
    "                        action = f'REMOVE;{row['ID_remove']};{row['ID_add']}'\n",
    "                    else:\n",
    "                        action = f'COMPARE'\n",
    "        actions.append(action)\n",
    "    \n",
    "    print(f'{set(sub_df['gene_add'])}\\t{set(actions)}')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mjol",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
